File: code-library/08-export-tables/export-tables.R
# ==============================================================================
#  MODULE 8: EXPORTING TABLES TO WORD, LATEX, AND EXCEL
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Purpose: Automate production of publication-ready tables from R output.
#  Primary tool: modelsummary (regression + descriptive tables to any format)
#  Also covers: gt, flextable, writexl for additional output types.
#
#  install.packages(c("modelsummary", "gt", "flextable", "writexl", "officer"))
# ==============================================================================

library(tidyverse)
library(modelsummary)   # regression tables to Word/LaTeX/HTML/Excel
library(gt)             # elegant HTML/Word tables from tibbles
library(flextable)      # Word-friendly tables
library(writexl)        # write Excel without Java dependency
library(broom)
library(sandwich)

ces <- readRDS("CES2020_clean.rds")


# ==============================================================================
# SECTION 1: REGRESSION TABLES WITH MODELSUMMARY
# ==============================================================================

# Workflow:
#   1. Fit models and put them in a named list
#   2. Call modelsummary() — output format is controlled by the output= argument
#      "default"       → console
#      "table.docx"    → Word
#      "table.tex"     → LaTeX
#      "table.html"    → HTML
#      "table.xlsx"    → Excel

m1 <- lm(imm_restrict ~ education, data = ces)
m2 <- lm(imm_restrict ~ education + age + factor(sex) + factor(census_region), data = ces)
m3 <- lm(imm_restrict ~ education + age + factor(sex) + factor(census_region) +
            ideology5 + party_id7, data = ces)

models <- list(
  "Bivariate"    = m1,
  "Demographics" = m2,
  "Full"         = m3
)

# Console display
modelsummary(
  models,
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  gof_map = c("nobs", "r.squared", "adj.r.squared", "AIC", "BIC"),
  title   = "OLS Regression: Predictors of Immigration Restrictionism"
)

# Export to Word
modelsummary(
  models,
  output  = "table_ols_restrict.docx",
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  gof_map = c("nobs", "r.squared", "adj.r.squared"),
  title   = "OLS Regression: Predictors of Immigration Restrictionism"
)

# Export to LaTeX (booktabs format for academic journals)
modelsummary(
  models,
  output   = "table_ols_restrict.tex",
  stars    = c("*" = .05, "**" = .01, "***" = .001),
  booktabs = TRUE,
  gof_map  = c("nobs", "r.squared", "adj.r.squared"),
  title    = "OLS Regression: Predictors of Immigration Restrictionism"
)


# ==============================================================================
# SECTION 2: LOGIT TABLE WITH ODDS RATIOS
# ==============================================================================

l1 <- glm(voted ~ education + age + factor(sex) + ideology5,
          data = ces, family = binomial())
l2 <- glm(voted ~ education + age + factor(sex) + ideology5 + party_id7 + college,
          data = ces, family = binomial())
l3 <- glm(voted ~ education + age + factor(sex) + ideology5 + party_id7 +
            college + white_nh + imm_restrict,
          data = ces, family = binomial())

logit_models <- list("Model 1" = l1, "Model 2" = l2, "Model 3" = l3)

modelsummary(
  logit_models,
  exponentiate = TRUE,   # display ORs instead of log-odds
  stars        = c("*" = .05, "**" = .01, "***" = .001),
  gof_map      = c("nobs", "logLik", "AIC"),
  title        = "Logistic Regression (Odds Ratios): Voter Turnout 2020"
)

modelsummary(
  logit_models,
  output       = "table_logit_turnout.docx",
  exponentiate = TRUE,
  stars        = c("*" = .05, "**" = .01, "***" = .001),
  gof_map      = c("nobs", "AIC"),
  title        = "Logistic Regression (Odds Ratios): Voter Turnout 2020"
)


# ==============================================================================
# SECTION 3: REGRESSION TABLE WITH ROBUST SEs
# ==============================================================================

# Pass a vcov argument to apply robust SEs to all models automatically
modelsummary(
  models,
  vcov    = "HC3",
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  gof_map = c("nobs", "r.squared", "adj.r.squared"),
  title   = "OLS with HC3 Robust Standard Errors"
)

# Clustered SEs by state
modelsummary(
  list("Clustered" = m3),
  vcov    = ~ state_fips,   # cluster on state_fips variable
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  title   = "OLS with State-Clustered Standard Errors"
)


# ==============================================================================
# SECTION 4: DESCRIPTIVE STATISTICS TABLE (TABLE 1)
# ==============================================================================

# datasummary_skim() from modelsummary gives a quick descriptive table
datasummary_skim(
  ces |> select(age, education, ideology5, imm_restrict, econ_retro,
                self_rated_health, college, voted, biden_voter, white_nh),
  title = "Table 1. Descriptive Statistics — CES 2020"
)

# Export to Word
datasummary_skim(
  ces |> select(age, education, ideology5, imm_restrict, econ_retro,
                college, voted, white_nh),
  output = "table1_descriptives.docx",
  title  = "Table 1. Descriptive Statistics"
)

# Custom Table 1 with specific statistics using datasummary()
datasummary(
  age + education + ideology5 + imm_restrict + econ_retro ~
    Mean + SD + Median + Min + Max + N,
  data  = ces,
  title = "Table 1. Descriptive Statistics"
)


# ==============================================================================
# SECTION 5: CROSSTAB TABLE WITH DATASUMMARY_CROSSTAB
# ==============================================================================

# datasummary_crosstab() produces a formatted two-way table
datasummary_crosstab(
  party_id3 ~ biden_voter,
  data  = filter(ces, voted == 1, !is.na(party_id3), !is.na(biden_voter)),
  title = "Vote Choice by Party Identification (Voters Only)"
)


# ==============================================================================
# SECTION 6: MEAN COMPARISON TABLE (T-TEST STYLE TABLE 2)
# ==============================================================================

# Build a group comparison table manually using dplyr, then format with gt

compare_df <- ces |>
  filter(!is.na(biden_voter)) |>
  group_by(biden_voter) |>
  summarise(
    mean_age       = mean(age, na.rm = TRUE),
    mean_education = mean(education, na.rm = TRUE),
    mean_ideology  = mean(ideology5, na.rm = TRUE),
    mean_restrict  = mean(imm_restrict, na.rm = TRUE),
    mean_econ      = mean(econ_retro, na.rm = TRUE),
    n              = n(),
    .groups = "drop"
  ) |>
  mutate(biden_voter = if_else(biden_voter == 1, "Biden Voters", "Trump Voters")) |>
  pivot_longer(-c(biden_voter, n),
               names_to  = "variable",
               values_to = "mean") |>
  pivot_wider(names_from = biden_voter, values_from = mean)

compare_df |>
  gt() |>
  tab_header(title = "Table 2. Mean Differences by Presidential Vote Choice") |>
  fmt_number(columns = where(is.numeric), decimals = 2) |>
  cols_label(variable = "Variable")


# ==============================================================================
# SECTION 7: FORMATTED TABLES WITH GT
# ==============================================================================

# gt produces high-quality HTML and Word tables from any tibble

summary_stats <- ces |>
  select(age, education, ideology5, imm_restrict, econ_retro) |>
  summarise(across(everything(),
                   list(Mean = ~ round(mean(.x, na.rm=TRUE), 2),
                        SD   = ~ round(sd(.x,   na.rm=TRUE), 2),
                        N    = ~ sum(!is.na(.x))),
                   .names = "{.col}__{.fn}")) |>
  pivot_longer(everything(),
               names_to  = c("Variable", ".value"),
               names_sep = "__")

summary_stats |>
  gt() |>
  tab_header(
    title    = "Descriptive Statistics",
    subtitle = "CES 2020 Common Content"
  ) |>
  tab_spanner(label = "Statistics", columns = c(Mean, SD, N)) |>
  fmt_number(columns = c(Mean, SD), decimals = 2) |>
  fmt_integer(columns = N) |>
  cols_align(align = "center", columns = c(Mean, SD, N)) |>
  tab_source_note("Source: CES 2020 (Schaffner, Ansolabehere, Luks 2021)")

# Save gt table as Word document
# summary_stats |> gt() |> ... |> gtsave("gt_descriptives.docx")


# ==============================================================================
# SECTION 8: EXPORT TO EXCEL WITH WRITEXL
# ==============================================================================

# writexl::write_xlsx() exports a list of dataframes as separate sheets

coef_table <- tidy(m3, conf.int = TRUE) |>
  mutate(across(where(is.numeric), ~ round(.x, 4)))

desc_table <- ces |>
  select(age, education, ideology5, imm_restrict, econ_retro) |>
  summarise(across(everything(),
                   list(mean = ~ mean(.x, na.rm=TRUE),
                        sd   = ~ sd(.x, na.rm=TRUE)),
                   .names = "{.col}_{.fn}"))

write_xlsx(
  list(
    "Regression Coefficients" = coef_table,
    "Descriptive Statistics"  = as.data.frame(desc_table)
  ),
  path = "ces_analysis_output.xlsx"
)

message("Excel file written: ces_analysis_output.xlsx")


# ==============================================================================
# SECTION 9: CUSTOM COEFFICIENT RENAME MAP
# ==============================================================================

# modelsummary lets you rename coefficient labels for cleaner table display

coef_labels <- c(
  "(Intercept)"           = "Intercept",
  "education"             = "Education (1–6)",
  "age"                   = "Age",
  "factor(sex)2"          = "Female (ref: Male)",
  "factor(census_region)2" = "Midwest (ref: Northeast)",
  "factor(census_region)3" = "South",
  "factor(census_region)4" = "West",
  "ideology5"             = "Ideology (1=Lib, 5=Con)",
  "party_id7"             = "Party ID (1=Strong Dem, 7=Strong Rep)"
)

modelsummary(
  list("Full OLS" = m3),
  coef_map = coef_labels,
  stars    = c("*" = .05, "**" = .01, "***" = .001),
  gof_map  = c("nobs", "r.squared"),
  title    = "OLS Regression with Custom Coefficient Labels"
)


# ==============================================================================
# SECTION 10: ADDING NOTES AND FORMATTING FOOTNOTES
# ==============================================================================

# modelsummary supports notes= for table footnotes

modelsummary(
  logit_models,
  exponentiate = TRUE,
  stars        = c("*" = .05, "**" = .01, "***" = .001),
  gof_map      = c("nobs", "AIC"),
  notes        = list(
    "Robust (HC3) standard errors in parentheses.",
    "Source: CES 2020 (Schaffner, Ansolabehere & Luks 2021).",
    "Reference categories: Female=Male, Region=Northeast."
  ),
  title = "Table 3. Logistic Regression: Voter Turnout"
)

message("Module 8 complete.")

File: code-library/08-export-tables/export-tables.py
# ==============================================================================
#  MODULE 8: EXPORTING TABLES TO EXCEL, WORD, AND LATEX
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Stack: pandas (Excel/CSV), openpyxl (Excel formatting),
#         tabulate (LaTeX/Markdown), python-docx (Word)
#
#  pip install openpyxl tabulate python-docx
# ==============================================================================

import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from tabulate import tabulate

ces = pd.read_parquet("CES2020_clean.parquet")


# ==============================================================================
# SECTION 1: BUILD A REGRESSION COEFFICIENT TABLE
# ==============================================================================

# Run three nested models
m1 = smf.ols("imm_restrict ~ education", data=ces).fit()
m2 = smf.ols("imm_restrict ~ education + age + C(sex) + C(census_region)", data=ces).fit()
m3 = smf.ols(
    "imm_restrict ~ education + age + C(sex) + C(census_region) + ideology5 + party_id7",
    data=ces
).fit()

models = {"Bivariate": m1, "Demographics": m2, "Full": m3}


def stars(p):
    if p < 0.001: return "***"
    if p < 0.01:  return "**"
    if p < 0.05:  return "*"
    return ""


def build_coef_table(models_dict):
    """Build a publication-style coefficient table from a dict of fitted models."""
    all_params = sorted(
        set(p for m in models_dict.values() for p in m.params.index),
        key=lambda x: (x == "Intercept", x)
    )

    rows = []
    for param in all_params:
        row = {"term": param}
        for name, m in models_dict.items():
            if param in m.params:
                b  = m.params[param]
                se = m.bse[param]
                p  = m.pvalues[param]
                row[name]          = f"{b:.3f}{stars(p)}"
                row[f"{name}_se"]  = f"({se:.3f})"
            else:
                row[name]         = ""
                row[f"{name}_se"] = ""
        rows.append(row)

    # Add model-level stats
    for stat, fn in [("N", lambda m: int(m.nobs)),
                     ("R²", lambda m: f"{m.rsquared:.3f}"),
                     ("Adj. R²", lambda m: f"{m.rsquared_adj:.3f}")]:
        row = {"term": stat}
        for name, m in models_dict.items():
            row[name]         = fn(m)
            row[f"{name}_se"] = ""
        rows.append(row)

    return pd.DataFrame(rows)


table = build_coef_table(models)
print(table.to_string(index=False))


# ==============================================================================
# SECTION 2: EXPORT TO CSV (UNIVERSAL)
# ==============================================================================

table.to_csv("table_ols_restrict.csv", index=False)
print("Saved: table_ols_restrict.csv")


# ==============================================================================
# SECTION 3: EXPORT TO EXCEL WITH FORMATTING
# ==============================================================================

# openpyxl allows cell-level formatting (bold headers, borders, column widths)

import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side

wb = openpyxl.Workbook()
ws = wb.active
ws.title = "OLS Regression"

# Write header
ws["A1"] = "OLS Regression: Predictors of Immigration Restrictionism"
ws["A1"].font = Font(bold=True, size=13)
ws.merge_cells("A1:G1")

# Column headers
headers = ["Term", "Bivariate", "(SE)", "Demographics", "(SE)", "Full", "(SE)"]
for col_idx, header in enumerate(headers, start=1):
    cell = ws.cell(row=3, column=col_idx, value=header)
    cell.font      = Font(bold=True)
    cell.alignment = Alignment(horizontal="center")
    cell.fill      = PatternFill(fill_type="solid", fgColor="DDEEFF")

# Write coefficient rows
thin_border = Border(
    bottom=Side(border_style="thin"),
    top=Side(border_style="thin")
)

col_order = ["term",
             "Bivariate", "Bivariate_se",
             "Demographics", "Demographics_se",
             "Full", "Full_se"]

for row_idx, (_, row) in enumerate(table.iterrows(), start=4):
    for col_idx, col_key in enumerate(col_order, start=1):
        cell = ws.cell(row=row_idx, column=col_idx,
                       value=row.get(col_key, ""))
        cell.alignment = Alignment(horizontal="center" if col_idx > 1 else "left")

# Auto-fit column widths (approximate)
for col in ws.columns:
    max_len = max((len(str(cell.value or "")) for cell in col), default=0)
    ws.column_dimensions[col[0].column_letter].width = max(max_len + 2, 10)

# Add a note row
note_row = ws.max_row + 1
ws.cell(row=note_row, column=1,
        value="Note: * p<.05, ** p<.01, *** p<.001. SEs in parentheses.")
ws.cell(row=note_row, column=1).font = Font(italic=True, size=9)

wb.save("table_ols_restrict.xlsx")
print("Saved: table_ols_restrict.xlsx")


# ==============================================================================
# SECTION 4: EXPORT TO WORD WITH PYTHON-DOCX
# ==============================================================================

from docx import Document
from docx.shared import Pt, Inches, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.table import WD_ALIGN_VERTICAL

doc = Document()

# Title
title = doc.add_heading("OLS Regression: Predictors of Immigration Restrictionism", level=1)
title.runs[0].font.size = Pt(14)

# Intro paragraph
n = int(m3.nobs)
r2 = m3.rsquared
doc.add_paragraph(
    f"Table 1 presents three nested OLS regression models predicting immigration "
    f"restrictionism. The full model (Model 3) uses N={n:,} observations and "
    f"explains {r2:.1%} of the variance in the outcome."
)

# Build table in docx
col_headers = ["Term", "Bivariate", "Demographics", "Full Model"]
display_cols = ["term", "Bivariate", "Demographics", "Full"]
se_cols      = ["", "Bivariate_se", "Demographics_se", "Full_se"]

word_table = doc.add_table(rows=1, cols=4)
word_table.style = "Table Grid"

# Header row
hdr_cells = word_table.rows[0].cells
for i, h in enumerate(col_headers):
    hdr_cells[i].text = h
    hdr_cells[i].paragraphs[0].runs[0].bold = True

# Data rows (coefficient + SE rows interleaved)
for _, row in table.iterrows():
    coef_row = word_table.add_row().cells
    for i, col in enumerate(display_cols):
        coef_row[i].text = str(row.get(col, ""))

    # SE row (if any SE exists for this row)
    has_se = any(str(row.get(s, "")).strip() not in ("", "nan")
                 for s in se_cols if s)
    if has_se:
        se_row = word_table.add_row().cells
        for i, col in enumerate(se_cols):
            se_row[i].text = str(row.get(col, "")) if col else ""
            se_row[i].paragraphs[0].runs[0].font.color.rgb = RGBColor(0x60, 0x60, 0x60)

# Note
note = doc.add_paragraph(
    "Note: * p<.05, ** p<.01, *** p<.001. Robust (HC3) standard errors in parentheses. "
    "Source: CES 2020 (Schaffner, Ansolabehere & Luks, 2021)."
)
note.runs[0].font.size = Pt(9)
note.runs[0].italic    = True

doc.save("table_ols_restrict.docx")
print("Saved: table_ols_restrict.docx")


# ==============================================================================
# SECTION 5: EXPORT TO LATEX
# ==============================================================================

def to_latex_table(models_dict, caption="OLS Regression Results", label="tab:ols"):
    """Build a LaTeX booktabs regression table."""
    col_names = list(models_dict.keys())
    all_params = []
    for name, m in models_dict.items():
        for p in m.params.index:
            if p not in all_params:
                all_params.append(p)

    lines = []
    lines.append(r"\begin{table}[htbp]")
    lines.append(r"\centering")
    lines.append(rf"\caption{{{caption}}}")
    lines.append(rf"\label{{{label}}}")
    lines.append(r"\begin{tabular}{l" + "c" * len(col_names) + "}")
    lines.append(r"\toprule")
    lines.append(" & " + " & ".join(col_names) + r" \\")
    lines.append(r"\midrule")

    for param in all_params:
        row_coef = [param.replace("_", r"\_")]
        row_se   = [""]
        for name, m in models_dict.items():
            if param in m.params:
                b  = m.params[param]
                se = m.bse[param]
                p  = m.pvalues[param]
                row_coef.append(f"{b:.3f}{stars(p)}")
                row_se.append(f"({se:.3f})")
            else:
                row_coef.append("")
                row_se.append("")
        lines.append(" & ".join(row_coef) + r" \\")
        lines.append(" & ".join(row_se)   + r" \\")

    lines.append(r"\midrule")
    # GOF stats
    for stat, fn in [("N", lambda m: str(int(m.nobs))),
                     ("$R^2$", lambda m: f"{m.rsquared:.3f}"),
                     ("Adj.~$R^2$", lambda m: f"{m.rsquared_adj:.3f}")]:
        row = [stat] + [fn(m) for m in models_dict.values()]
        lines.append(" & ".join(row) + r" \\")

    lines.append(r"\bottomrule")
    lines.append(
        r"\multicolumn{" + str(len(col_names) + 1) + r"}{l}{"
        r"\textit{Note:} $^{*}p<.05$, $^{**}p<.01$, $^{***}p<.001$. "
        r"SEs in parentheses.} \\"
    )
    lines.append(r"\end{tabular}")
    lines.append(r"\end{table}")

    return "\n".join(lines)


latex_output = to_latex_table(models, caption="OLS: Immigration Restrictionism")

with open("table_ols_restrict.tex", "w") as f:
    f.write(latex_output)

print("Saved: table_ols_restrict.tex")


# ==============================================================================
# SECTION 6: DESCRIPTIVE STATISTICS TABLE
# ==============================================================================

desc_vars = ["age", "education", "ideology5", "imm_restrict",
             "econ_retro", "self_rated_health", "college", "voted",
             "biden_voter", "white_nh"]

desc = (
    ces[desc_vars]
    .agg(["mean", "std", "median", "min", "max", lambda x: x.notna().sum()])
    .T
    .rename(columns={"<lambda_0>": "N"})
    .round(3)
)

desc.to_csv("table1_descriptives.csv")
print("\nTable 1 (descriptives):")
print(tabulate(desc, headers="keys", tablefmt="pipe", floatfmt=".3f"))


# ==============================================================================
# SECTION 7: LOGIT TABLE WITH ODDS RATIOS
# ==============================================================================

import numpy as np

l1 = smf.logit("voted ~ education + age + C(sex) + ideology5", data=ces).fit(disp=False)
l2 = smf.logit("voted ~ education + age + C(sex) + ideology5 + party_id7 + college", data=ces).fit(disp=False)
l3 = smf.logit("voted ~ education + age + C(sex) + ideology5 + party_id7 + college + white_nh + imm_restrict", data=ces).fit(disp=False)

logit_models = {"Model 1": l1, "Model 2": l2, "Model 3": l3}

def build_or_table(models_dict):
    """Build an odds-ratio table from logit models."""
    all_params = []
    for m in models_dict.values():
        for p in m.params.index:
            if p not in all_params:
                all_params.append(p)

    rows = []
    for param in all_params:
        row = {"term": param}
        for name, m in models_dict.items():
            if param in m.params:
                OR = np.exp(m.params[param])
                se = m.bse[param]
                p  = m.pvalues[param]
                ci_lo = np.exp(m.conf_int().loc[param, 0])
                ci_hi = np.exp(m.conf_int().loc[param, 1])
                row[name] = f"{OR:.3f}{stars(p)}"
                row[f"{name}_ci"] = f"[{ci_lo:.3f}, {ci_hi:.3f}]"
            else:
                row[name]         = ""
                row[f"{name}_ci"] = ""
        rows.append(row)

    for stat, fn in [("N", lambda m: int(m.nobs)),
                     ("AIC", lambda m: f"{m.aic:.1f}"),
                     ("Pseudo R²", lambda m: f"{1 - m.llf / smf.logit('voted~1', data=ces).fit(disp=False).llf:.3f}")]:
        row = {"term": stat}
        for name, m in models_dict.items():
            row[name]         = fn(m)
            row[f"{name}_ci"] = ""
        rows.append(row)

    return pd.DataFrame(rows)

or_table = build_or_table(logit_models)
or_table.to_csv("table_logit_turnout.csv", index=False)
print("\nLogit OR Table saved: table_logit_turnout.csv")


# ==============================================================================
# SECTION 8: EXPORT MULTIPLE SHEETS TO EXCEL
# ==============================================================================

# pd.ExcelWriter allows writing multiple DataFrames to separate sheets
with pd.ExcelWriter("ces_analysis_output.xlsx", engine="openpyxl") as writer:
    desc.to_excel(writer, sheet_name="Descriptive Statistics")
    table.to_excel(writer, sheet_name="OLS Regression", index=False)
    or_table.to_excel(writer, sheet_name="Logit ORs", index=False)

print("Saved: ces_analysis_output.xlsx (3 sheets)")


# ==============================================================================
# SECTION 9: MARKDOWN TABLE (FOR GITHUB README OR NOTEBOOKS)
# ==============================================================================

print("\nMarkdown table (for README or notebook):")
print(tabulate(
    desc.head(6),
    headers="keys",
    tablefmt="github",
    floatfmt=".2f"
))

print("\nModule 8 complete.")
